dial/pipeline.go

```diff
@@ -0,0 +1,313 @@
+}
+
+// Values returns the connection and error fields from this request.
+// Both return values may be nil or incoherent unless the request has completed (see IsComplete).
+func (req *Request) Values() (inet.Conn, error) {
+	return req.conn, req.err
+}
+
+// AddCallback adds a function that will be invoked when this request completes, either in success
+// or in failure.
+func (req *Request) AddCallback(cb func()) {
+	req.callbacks = append(req.callbacks, cb)
+}
+
+// Job represents a dial attempt to a single multiaddr. It is associated 1-* to a Request.
+// It can carry its own context and its own callbacks.
+type Job struct {
+	status int32
+	req    *Request
+
+	ctx  context.Context
+	addr ma.Multiaddr
+
+	callbacks []func()
+
+	// result.
+	tconn tpt.Conn
+	err   error
+
+	// Channel where completed dials should be sent.
+	completeCh chan *Job
+}
+
+func NewDialJob(ctx context.Context, req *Request, addr ma.Multiaddr) *Job {
+	return &Job{req: req, ctx: ctx, addr: addr, status: jobInflight}
+}
+
+func (j *Job) Cancelled() bool {
+	select {
+	case <-j.req.ctx.Done():
+		return true
+	case <-j.ctx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+func (j *Job) AddCallback(cb func()) {
+	j.callbacks = append(j.callbacks, cb)
+}
+
+func (j *Job) Complete(conn tpt.Conn, err error) error {
+	if !atomic.CompareAndSwapInt32(&j.status, jobInflight, jobFinished) {
+		return ErrDialJobInvalidStatus
+	}
+
+	j.tconn, j.err = conn, err
+	for i := len(j.callbacks); i > 0; i-- {
+		j.callbacks[i-1]()
+	}
+
+	j.completeCh <- j
+	return nil
+}
+
+type dialJobs []*Job
+
+func (djs *dialJobs) sift() (success dialJobs, failed dialJobs) {
+	for _, dj := range *djs {
+		if dj.err == nil {
+			success = append(success, dj)
+			continue
+		}
+		failed = append(failed, dj)
+	}
+	return success, failed
+}
+
+type AddConnFn func(tc tpt.Conn, dir inet.Direction) (inet.Conn, error)
+
+type Pipeline struct {
+	lk  sync.RWMutex
+	ctx context.Context
+
+	net inet.Network
+
+	reqPreparers []RequestPreparer
+	planner      Planner
+	throttler    Throttler
+	jobPreparers []JobPreparer
+	executor     Executor
+	selector     Selector
+
+	throttleCh chan *Job
+	dialCh     chan *Job
+
+	addConnFn AddConnFn
+}
+
+func (p *Pipeline) Component(name string, comp interface{}) error {
+	switch comp.(type) {
+	case RequestPreparer:
+		p.reqPreparers = append(p.reqPreparers, comp.(RequestPreparer))
+	case Planner:
+		p.planner = comp.(Planner)
+	case Throttler:
+		p.throttler = comp.(Throttler)
+	case JobPreparer:
+		p.jobPreparers = append(p.jobPreparers, comp.(JobPreparer))
+	case Executor:
+		p.executor = comp.(Executor)
+	case Selector:
+		p.selector = comp.(Selector)
+	default:
+		return errors.New("unrecognized pipeline component")
+	}
+	return nil
+}
+
+func NewPipeline(ctx context.Context, net inet.Network, addConnFn AddConnFn) *Pipeline {
+	pipeline := &Pipeline{
+		ctx:       ctx,
+		net:       net,
+		addConnFn: addConnFn,
```

Comment:
do we *need* this callback or is there some way we can completely separate this from the network? That is, it would be nice if the pipeline could be completely modular and *return* connections to the network instead of feeding them in like this.

Reply:
Especially because this makes it hard for users to define custom pipelines.

Reply:
Actually, you're right. The pipeline could simply return a connection, and the `Swarm` could add it locally. I like that much better.